docs/design/custom-plugin.md

```diff
@@ -0,0 +1,128 @@
+# Custom Plugin
+
+## Background
+
+Until now, plugins like `binpack`, `drf`, `gang` are provided by official.
+But some users may want to implement the plugin by themselves. So if scheduler can dynamically load plugins, that will make it more flexible when handling different business scenarios.
+
+## How to build a plugin
+
+### 1. Coding
+
+```go
+// magic.go
+
+package main  // note!!! package must be named main
+
+import (
+    "volcano.sh/volcano/pkg/scheduler/framework"
+)
+
+const PluginName = "magic"
+
+type magicPlugin struct {}
+
+func (mp *magicPlugin) Name() string {
+    return PluginName
+}
+
+func New(arguments framework.Arguments) framework.Plugin {  // `New` is PluginBuilder
+	return &magicPlugin{}
+}
+
+func (mp *magicPlugin) OnSessionOpen(ssn *framework.Session) {}
+
+func (mp *magicPlugin) OnSessionClose(ssn *framework.Session) {}
+```
+
+### 2. Build the plugin to .so
+
+#### A. Use musl-libc build plugin
+
+Because the default `vc-scheduler` base image is `alpine`, which only has `musl-libc`, so we should use `musl-gcc` to
+build the plugin.
+
+```bash
+docker run -v `pwd`:/work golang:1.14-alpine sh -c "cd /work && apk add musl-dev gcc && go build -buildmode=plugin magic.go"
+```
+
+Or build the plugin in local.
+
+```bash
+# install musl
+wget http://musl.libc.org/releases/musl-1.2.1.tar.gz
+tar -xf musl-1.2.1.tar.gz && cd musl-1.2.1
+./configure
+make && sudo make install
+
+# build plugin
+CC=/usr/local/musl/bin/musl-gcc CGO_ENABLED=1 go build -buildmode=plugin magic.go
+```
+
+#### B. Use gnu-libc build plugin
+
+If want to use `ubuntu` as base image, you can use `gnu-libc` to build the plugin. Since most Linux OS have `gnu-libc`,
+you can just build the plugin in local.
+
+```bash
+# default CC is gcc
+CGO_ENABLED=1 go build -buildmode=plugin magic.go
+```
+
+### 3. Add plugins into container
+
+Your can build your docker image
+
+```dockerfile
+FROM volcano.sh/vc-scheduler:${VERSION}
+
+COPY plugins plugins
+```
+
+Or just use `pvc` to mount these plugins
+
+### 4. Specify deployment
+```yaml
+...
+    containers:
+    - name: volcano-scheduler
+      image: volcano.sh/vc-scheduler:${VERSION}
+      args:
+        - --logtostderr
+        - --scheduler-conf=/volcano.scheduler/volcano-scheduler.conf
+        - -v=3
+        - --plugins-dir=plugins  # specify plugins dir path
```

Comment:
So official plugin will be built and loaded as privous way, and custom plugin can be built and loaded in this way.

Reply:
yes